pallets/profile/src/lib.rs

```diff
@@ -0,0 +1,277 @@
+        /// The entry type for the activity is not valid.
+        InvalidEntryTypeInput,
+        /// Rotation must be to new key, not to existing tied account.
+        CannotRotateToSameAccount,
+    }
+
+    #[pallet::event]
+    #[pallet::generate_deposit(pub(super) fn deposit_event)]
+    pub enum Event<T: Config> {
+        /// A new profile has been created.
+		/// \[Creator, Profile Identifier\]
+        ProfileSet { 
+            who: CreatorOf<T>, 
+            identifier: ProfileIdOf
+        },
+
+        /// Existing Profile's Key has been rotated.
+        /// \[Old Profile Holder, Profile Identifier, New public of the Holder]
+        KeyRotated {
+            who: CreatorOf<T>,
+            identifier: ProfileIdOf,
+            new_key: CreatorOf<T>,
+        }
+    }
+
+    #[pallet::call]
+    impl<T: Config> Pallet<T> {
+        /// Creates a new Profile with associated details.
+		///
+        /// This extrinsic creates a new Profile to be mapped to a CORD account.
+		/// Also it expects a vector of Data Keys and its values. 
+        /// This HashMap is tied to the generated Profile Identifier.
+        /// But currently expects the user given Data Keys to start with `pub_` prefix only.
+        ///
+		/// # Parameters
+		/// - `origin`: The origin of the call, which must be signed by the creator of the profile.
+		/// - `data`: The data is a vector of a HashMap of the Data Key and associated Data Values.
+		///
+		/// # Returns
+		/// Returns `Ok(())` if the Profile is succesfully been created.
+		/// or an `Err` if the operation fails due to a same profile already existing.
+		///
+		/// # Errors
+        /// - `InvalidIdentifierLength`: If the newly creted Profile Identifier exceeds limit.
+		/// - `ProfileAlreadyExists`: If the newly created Profile already exists. 
+        ///   This happens if a user creates a duplicate profile with the same account key-pair.
+        /// - `InvalidKeyPrefix`: If the Profile Data Key starts with anything other than `pub_`.
+        #[pallet::call_index(0)]
+        #[pallet::weight({10_000})]
+        pub fn set_profile(
+            origin: OriginFor<T>,
+            data: Vec<(DataKeyOf<T>, DataValueOf<T>)>,
+        ) -> DispatchResult {
+            let who = ensure_signed(origin)?;
+            let digest = T::Hashing::hash(&who.encode());
+            let pallet_name = <Self as frame_support::traits::PalletInfoAccess>::name();
+            let profile_id = <cord_uri::Pallet<T> as Identifier>::build(&(digest).encode()[..], pallet_name)
+                .map_err(|_| Error::<T>::InvalidIdentifierLength)?;
+
+            ensure!(
+                !Profiles::<T>::contains_key(&profile_id),
+                Error::<T>::ProfileAlreadyExists
+            );
+
+            let profile = ProfileMetadataOf::<T> { latest_key: who.clone() }; 
+            Profiles::<T>::insert(&profile_id, profile);
+            AccountProfiles::<T>::insert(&who, &profile_id);
+
```

Comment:
most of the times, lets insert the Storage entries after all validations happen. What if this `set_profile()` fails with `InvalidKeyPrefix` below. In that case, we will have many dangling profile identifier, which are not valid.